### DISCUSSION TITLE
openai- fetch error

### DISCUSSION THREAD

**Discussion by Moukarnold** [neuefische/questions#624] [Web]
hello,
i have a form composed of 3 select inputs and a submit button. the submit button generates a fetch via an open ai. the values selected in the select buttons are used as parameters for the fetch. When i submit
i become this error. so can someone helps?

![Bildschirm­foto 2023-06-02 um 12 41 09](https://github.com/neuefische/questions/assets/128094305/e1f8ed2e-c578-4dc8-99ff-be7a171956ae)

https://github.com/Moukarnold/capstone-project/tree/openai/pages

**Comment by onemanwenttomow** [comment]
Hi @Moukarnold 

I am not particularly familiar with the openai api, but here are my suggestions. 

I think that the messages part of the object that you pass to `createChatCompletion` does not seem to match up with [what the documentation is asking for](https://platform.openai.com/docs/guides/chat/introduction). 

This was your old setup -> [see here](https://github.com/Moukarnold/capstone-project/blob/openai/utils/openaiConfig.js)

This is what i would suggest changing it to. 

``` js
messages: [
  {
    role: "system",
    content:
      "You can only respond to concepts that exist in your life. You can only respond to topics that correspond to your expertise",
  },
  {
    role: "user",
    content: `Please generate a tongue twister in the language ${prompt.Languages}
  - of type ${prompt.options} and complexity level ${prompt.Levels}.
  Please provide a maximum of 2 sentences.`,
  },
],
```

I think you also need to change what you return from the function. 

``` js
return response.data.choices[0].message;  // instead of response.data.choices[0].messages;
```

And finally, in the component that receives the response, i think you would need to update the jsx to 

``` js
{loading ? <p>loading..</p> : <p>{answer?.content}</p>}
```

I would also consider wrapping the call to `openai.createChatCompletion` in a `try catch` block. That way you could potentially find about more information about the error if something goes wrong.

**Comment by Satttoshi** [comment]
Hey. Your gpt file and `CreateChatCompletion`-api request seems fine to me. Is this discussion resolved?

500 Server Errors on fetching Openai could happen due to trying to access to the GPT-3.5-turbo model, which is a paid per fetch model. You need to have a credit card linked to your account which inherits your used `OPENAI_API_KEY`. Single fetches on the GPT-3.5-turbo model are rather cheap, you could link your credit card and set a monthly fetching limit to like 5 dollars. Then you would need about thousand fetches to reach that.

If you still want a free tier solution, then you might consider using the `createCompletion` Function models like DaVinci. But you will lose the ability to do system instructions then.

I checked where you fetch, how you handle your data and how. Your code should work with a paid plan. I assume that a `.env.local` exists on your local repository. If you use Vercel you might run into the issue trying to run a `serverless function` there, which has a fixed duration to run and timeouts if exceeded. Try to debug locally first, before trying to run it on a deployment Server.
